src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```
@@ -0,0 +1,531 @@
+            var response = AwsHttpClientHelper.GetECSResponse(relativeUri).GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2Response().GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case "AWS_SESSION_TOKEN":
+                            sessionToken = (string)pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
```

Not sure we should be checking for invalid properties here.

Shouldn't we just be looking for the token and ignoring the rest.

If so, this entire method could be simplified to:

    return properties?.FirstOrDefault(x => x.Key.Equals("AWS_SESSION_TOKEN", StringComparison.OrdinalIgnoreCase)).Value;

---

I was thinking that we shouldn't silently ignore additional mechanism parameters if user has specified them. As a user I would rather expect a system to throw exception if I specified incorrect or inexistent option.
Moreover I was following existing implementation: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs#L151)